File: scala/sdktest/MoonEntityTest.scala
// Generated basic-flow test for the moon entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped MoonTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.solardemosdk.core.{Helpers, SdkEntity, SolardemoSDK}
import voxgig.solardemosdk.utility.struct.Struct

object MoonEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("moon.instance") {
      val testsdk = SolardemoSDK.testSDK()
      val ent = testsdk.moon(null)
      rep.check("moon.instance", ent != null, "expected non-null moon entity")
    }

    rep.scope("moon.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/moon/MoonTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = SolardemoSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("moon01", "MOON01")
      idmap.put("moon02", "MOON02")
      idmap.put("moon03", "MOON03")
      idmap.put("planet01", "PLANET01")
      idmap.put("planet02", "PLANET02")
      idmap.put("planet03", "PLANET03")
      idmap.put("planet_id", "PLANET01")
      val now = System.currentTimeMillis()

      // CREATE
      val moonRef01Ent = client.moon(null)
      var moonRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.moon"), "moon_ref01"))
      moonRef01Data.put("planet_id", idmap.get("planet01"))
      val moonRef01DataResult = moonRef01Ent.create(moonRef01Data, null)
      moonRef01Data = Helpers.toMapAny(moonRef01DataResult match { case e: SdkEntity => e.data(); case o => o })
      rep.check("moon.create.map", moonRef01Data != null, "expected create result to be a map")
      rep.check("moon.create.id", moonRef01Data != null && moonRef01Data.get("id") != null, "expected created entity to have an id")

      // LIST
      val moonRef01Match = new LinkedHashMap[String, Object]()
      moonRef01Match.put("planet_id", idmap.get("planet01"))
      val moonRef01ListResult = moonRef01Ent.list(moonRef01Match, null)
      rep.check("moon.list.islist", moonRef01ListResult.isInstanceOf[JList[?]], "expected list result to be an array, got " + moonRef01ListResult)
      val moonRef01List = moonRef01ListResult.asInstanceOf[JList[Object]]
      val moonRef01ListFound = Struct.select(
          SdkTestSupport.entityListToData(moonRef01List), SdkTestSupport.om("id" -> moonRef01Data.get("id")))
      rep.check("moon.list.exists", !Struct.isempty(moonRef01ListFound), "expected to find created entity in list")

      // UPDATE
      val moonRef01DataUp0Up = new LinkedHashMap[String, Object]()
      moonRef01DataUp0Up.put("id", moonRef01Data.get("id"))
      moonRef01DataUp0Up.put("planet_id", idmap.get("planet_id"))
      val moonRef01MarkdefUp0Name = "kind"
      val moonRef01MarkdefUp0Value = "Mark01-moon_ref01_" + now
      moonRef01DataUp0Up.put(moonRef01MarkdefUp0Name, moonRef01MarkdefUp0Value)
      val moonRef01ResdataUp0Result = moonRef01Ent.update(moonRef01DataUp0Up, null)
      val moonRef01ResdataUp0 = Helpers.toMapAny(moonRef01ResdataUp0Result match { case e: SdkEntity => e.data(); case o => o })
      rep.check("moon.update.map", moonRef01ResdataUp0 != null, "expected update result to be a map")
      rep.eq("moon.update.id", moonRef01DataUp0Up.get("id"), moonRef01ResdataUp0.get("id"))
      rep.eq("moon.update.mark", moonRef01MarkdefUp0Value, moonRef01ResdataUp0.get(moonRef01MarkdefUp0Name))

      // LOAD
      val moonRef01MatchDt0 = new LinkedHashMap[String, Object]()
      moonRef01MatchDt0.put("id", moonRef01Data.get("id"))
      val moonRef01DataDt0Loaded = moonRef01Ent.load(moonRef01MatchDt0, null)
      val moonRef01DataDt0LoadResult = Helpers.toMapAny(moonRef01DataDt0Loaded match { case e: SdkEntity => e.data(); case o => o })
      rep.check("moon.load.map", moonRef01DataDt0LoadResult != null, "expected load result to be a map")
      rep.eq("moon.load.id", moonRef01Data.get("id"), moonRef01DataDt0LoadResult.get("id"))

      // REMOVE
      val moonRef01MatchRm0 = new LinkedHashMap[String, Object]()
      moonRef01MatchRm0.put("id", moonRef01Data.get("id"))
      moonRef01Ent.remove(moonRef01MatchRm0, null)

      // LIST
      val moonRef01MatchRt0 = new LinkedHashMap[String, Object]()
      moonRef01MatchRt0.put("planet_id", idmap.get("planet01"))
      val moonRef01ListRt0Result = moonRef01Ent.list(moonRef01MatchRt0, null)
      rep.check("moon.list.islist", moonRef01ListRt0Result.isInstanceOf[JList[?]], "expected list result to be an array, got " + moonRef01ListRt0Result)
      val moonRef01ListRt0 = moonRef01ListRt0Result.asInstanceOf[JList[Object]]
      val moonRef01ListRt0NotFound = Struct.select(
          SdkTestSupport.entityListToData(moonRef01ListRt0), SdkTestSupport.om("id" -> moonRef01Data.get("id")))
      rep.check("moon.list.notexists", Struct.isempty(moonRef01ListRt0NotFound), "expected removed entity to not be in list")
    }
  }
}

File: rust/tests/moon_entity_test.rs
// Generated basic-flow test for the moon entity (model-driven;
// mirrors the go TestEntity generator).

#![allow(unused_variables, unused_mut, unused_imports)]

mod common;

use std::rc::Rc;

use common::*;

use solardemo_sdk::core::helpers::{getp, getpath, ja, jo, now_ms, setp, to_map};
use solardemo_sdk::utility::voxgigstruct as vs;
use solardemo_sdk::{test_sdk, Entity, SolardemoEntity, SolardemoSDK, Value};

#[test]
fn moon_entity_instance() {
    let testsdk = test_sdk(Value::Noval, Value::Noval);
    let ent = testsdk.moon(Value::Noval);
    assert_eq!(ent.get_name(), "moon");
}

#[test]
fn moon_entity_stream() {
    // stream() runs the list op through the full pipeline and yields each
    // result item. Seed two entities via test mode; with the `streaming`
    // feature active it yields the feature's incremental items, else it
    // falls back to the materialised items — either way every item yields.
    let seed = jo(vec![(
        "entity",
        jo(vec![(
            "moon",
            jo(vec![
                ("strm01", jo(vec![("id", Value::str("strm01"))])),
                ("strm02", jo(vec![("id", Value::str("strm02"))])),
            ]),
        )]),
    )]);

    let sdkopts = jo(vec![(
        "feature",
        jo(vec![("streaming", jo(vec![("active", Value::Bool(true))]))]),
    )]);

    let testsdk = test_sdk(seed.clone(), sdkopts);
    let ent = testsdk.moon(Value::Noval);
    let items: Vec<Value> = ent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(items.len(), 2, "stream should yield both seeded items");

    // Fallback: streaming inactive still yields both materialised items.
    let plainsdk = test_sdk(seed, Value::Noval);
    let plainent = plainsdk.moon(Value::Noval);
    let plain_items: Vec<Value> = plainent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(plain_items.len(), 2, "fallback stream should yield both items");
}

#[test]
fn moon_entity_basic() {
    let setup = moon_basic_setup(Value::Noval);
    // Per-op sdk-test-control.json skip — the basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    let mode = if setup.live { "live" } else { "unit" };
    for op in ["create", "list", "update", "load", "remove"] {
        let (skip, reason) = is_control_skipped("entityOp", &format!("moon.{}", op), mode);
        if skip {
            let reason = if reason.is_empty() {
                "skipped via sdk-test-control.json".to_string()
            } else {
                reason
            };
            eprintln!("skip: {}", reason);
            return;
        }
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only {
        eprintln!("skip: live entity test uses synthetic IDs from fixture — set SOLARDEMO_TEST_MOON_ENTID JSON to run live");
        return;
    }
    let client = setup.client.clone();
    // CREATE
    let moon_ref01_ent = client.moon(Value::Noval);
    let moon_ref01_data = to_map(&getp(
        &getpath(&["new", "moon"], &setup.data),
        "moon_ref01",
    ));
    setp(&moon_ref01_data, "planet_id", getp(&setup.idmap, "planet01"));

    let moon_ref01_data_result = moon_ref01_ent
        .create(moon_ref01_data.clone(), Value::Noval)
        .expect("create failed");
    let moon_ref01_data = to_map(&moon_ref01_data_result.data(None));
    assert!(
        matches!(moon_ref01_data, Value::Map(_)),
        "expected create result to be a map"
    );
    assert!(
        !getp(&moon_ref01_data, "id").is_noval(),
        "expected created entity to have an id"
    );

    // LIST
    let moon_ref01_match = Value::empty_map();
    setp(&moon_ref01_match, "planet_id", getp(&setup.idmap, "planet01"));

    let moon_ref01_list = moon_ref01_ent
        .list(moon_ref01_match.clone(), Value::Noval)
        .expect("list failed");
    // list resolves to one ENTITY per record; the flow asserts on the
    // records, so map each through data().
    let moon_ref01_list = ja(moon_ref01_list.iter().map(|e| e.data(None)).collect::<Vec<Value>>());

    let found_item = vs::select(
        &entity_list_to_data(&moon_ref01_list),
        &jo(vec![("id", getp(&moon_ref01_data, "id"))]),
    );
    assert!(
        !vs::is_empty(&found_item),
        "expected to find created entity in list"
    );

    // UPDATE
    let moon_ref01_data_up0_up = Value::empty_map();
    setp(&moon_ref01_data_up0_up, "id", getp(&moon_ref01_data, "id"));
    setp(&moon_ref01_data_up0_up, "planet_id", getp(&setup.idmap, "planet_id"));

    let moon_ref01_markdef_up0_name = "kind";
    let moon_ref01_markdef_up0_value = format!("Mark01-moon_ref01_{}", setup.now);
    setp(
        &moon_ref01_data_up0_up,
        moon_ref01_markdef_up0_name,
        Value::str(moon_ref01_markdef_up0_value.clone()),
    );

    let moon_ref01_resdata_up0_result = moon_ref01_ent
        .update(moon_ref01_data_up0_up.clone(), Value::Noval)
        .expect("update failed");
    let moon_ref01_resdata_up0 = to_map(&moon_ref01_resdata_up0_result.data(None));
    assert!(
        matches!(moon_ref01_resdata_up0, Value::Map(_)),
        "expected update result to be a map"
    );
    assert_eq!(
        getp(&moon_ref01_resdata_up0, "id"),
        getp(&moon_ref01_data_up0_up, "id"),
        "expected update result id to match"
    );
    assert_eq!(
        getp(&moon_ref01_resdata_up0, moon_ref01_markdef_up0_name),
        Value::str(moon_ref01_markdef_up0_value.clone()),
        "expected {} to be updated",
        moon_ref01_markdef_up0_name
    );

    // LOAD
    let moon_ref01_match_dt0 = jo(vec![("id", getp(&moon_ref01_data, "id"))]);
    let moon_ref01_data_dt0_loaded = moon_ref01_ent
        .load(moon_ref01_match_dt0.clone(), Value::Noval)
        .expect("load failed");
    let moon_ref01_data_dt0_load_result = to_map(&moon_ref01_data_dt0_loaded.data(None));
    assert!(
        matches!(moon_ref01_data_dt0_load_result, Value::Map(_)),
        "expected load result to be a map"
    );
    assert_eq!(
        getp(&moon_ref01_data_dt0_load_result, "id"),
        getp(&moon_ref01_data, "id"),
        "expected load result id to match"
    );

    // REMOVE
    let moon_ref01_match_rm0 = jo(vec![("id", getp(&moon_ref01_data, "id"))]);
    moon_ref01_ent
        .remove(moon_ref01_match_rm0.clone(), Value::Noval)
        .expect("remove failed");

    // LIST
    let moon_ref01_match_rt0 = Value::empty_map();
    setp(&moon_ref01_match_rt0, "planet_id", getp(&setup.idmap, "planet01"));

    let moon_ref01_list_rt0 = moon_ref01_ent
        .list(moon_ref01_match_rt0.clone(), Value::Noval)
        .expect("list failed");
    // list resolves to one ENTITY per record; the flow asserts on the
    // records, so map each through data().
    let moon_ref01_list_rt0 = ja(moon_ref01_list_rt0.iter().map(|e| e.data(None)).collect::<Vec<Value>>());

    let not_found_item = vs::select(
        &entity_list_to_data(&moon_ref01_list_rt0),
        &jo(vec![("id", getp(&moon_ref01_data, "id"))]),
    );
    assert!(
        vs::is_empty(&not_found_item),
        "expected removed entity to not be in list"
    );

}

fn moon_basic_setup(extra: Value) -> EntityTestSetup {
    load_env_local();

    let mut entity_data_file = manifest_dir();
    entity_data_file.push("..");
    entity_data_file.push(".sdk");
    entity_data_file.push("test");
    entity_data_file.push("entity");
    entity_data_file.push("moon");
    entity_data_file.push("MoonTestData.json");

    let entity_data = read_json(&entity_data_file);

    let options = jo(vec![("entity", getp(&entity_data, "existing"))]);

    let client = test_sdk(options, extra.clone());

    // Generate idmap via transform, matching the TS pattern.
    let idmap = vs::transform(
        &ja(vec![Value::str("moon01"), Value::str("moon02"), Value::str("moon03"), Value::str("planet01"), Value::str("planet02"), Value::str("planet03")]),
        &jo(vec![(
            "`$PACK`",
            ja(vec![
                Value::str(""),
                jo(vec![
                    ("`$KEY`", Value::str("`$COPY`")),
                    (
                        "`$VAL`",
                        ja(vec![
                            Value::str("`$FORMAT`"),
                            Value::str("upper"),
                            Value::str("`$COPY`"),
                        ]),
                    ),
                ]),
            ]),
        )]),
        None,
    )
    .unwrap_or_else(|_| Value::empty_map());

    // Detect ENTID env override before env_override consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's.
    let entid_env_raw = std::env::var("SOLARDEMO_TEST_MOON_ENTID").unwrap_or_default();
    let idmap_overridden =
        !entid_env_raw.trim().is_empty() && entid_env_raw.trim().starts_with('{');

    let env = env_override(jo(vec![
        ("SOLARDEMO_TEST_MOON_ENTID", idmap.clone()),
        ("SOLARDEMO_TEST_LIVE", Value::str("FALSE")),
        ("SOLARDEMO_TEST_EXPLAIN", Value::str("FALSE")),
    ]));

    let idmap_resolved = match to_map(&getp(&env, "SOLARDEMO_TEST_MOON_ENTID")) {
        Value::Map(m) => Value::Map(m),
        _ => to_map(&idmap),
    };

    // Add planet_id alias for the update test.
    if getp(&idmap_resolved, "planet_id").is_noval() {
        let aliased = getp(&idmap_resolved, "planet01");
        setp(&idmap_resolved, "planet_id", aliased);
    }

    let live = getp(&env, "SOLARDEMO_TEST_LIVE") == Value::str("TRUE");

    let client = if live {
        let merged = vs::merge(
            &ja(vec![jo(vec![]), extra]),
            None,
        );
        SolardemoSDK::new(to_map(&merged))
    } else {
        client
    };

    EntityTestSetup {
        client,
        data: entity_data,
        idmap: idmap_resolved,
        env: env.clone(),
        explain: getp(&env, "SOLARDEMO_TEST_EXPLAIN") == Value::str("TRUE"),
        live,
        synthetic_only: live && !idmap_overridden,
        now: now_ms(),
    }
}
